### PR TITLE
Xrc 1218 port uvatlas to linux

### PR DESCRIPTION
**Pull request by kiran1228**
@svensht2 @DillonSkeehan wanted some eyes on to make sure there are no major red flags but everything seems to work.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    #    UVAtlas is too slow in debug so use optimization flags to speed up
+    string(REPLACE "/Od" "/O2" CMAKE_C_FLAGS_DEBUG ${CMAKE_C_FLAGS_DEBUG})
+    string(REPLACE "/Ob0" "" CMAKE_C_FLAGS_DEBUG ${CMAKE_C_FLAGS_DEBUG})
+    string(REPLACE "/RTC1" "" CMAKE_C_FLAGS_DEBUG ${CMAKE_C_FLAGS_DEBUG})
+    string(REPLACE "/Od" "/O2" CMAKE_CXX_FLAGS_DEBUG ${CMAKE_CXX_FLAGS_DEBUG})
+    string(REPLACE "/Ob0" "" CMAKE_CXX_FLAGS_DEBUG ${CMAKE_CXX_FLAGS_DEBUG})
+    string(REPLACE "/RTC1" "" CMAKE_CXX_FLAGS_DEBUG ${CMAKE_CXX_FLAGS_DEBUG})
+elseif(UNIX)
+    set(CXX_COMMON_FLAGS "-std=c++11 -fPIC")
```
Comment (DillonSkeehan) [inline CMakeLists.txt]: is it possible to only do this for this project, general `set` makes me nervous

File: UVAtlas/UVAtlas_2017_Win10.vcxproj [hunk 2]
```diff
@@ -132,7 +132,7 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
     <OutDir>Bin\Desktop_2017_Win10\$(Platform)\$(Configuration)\</OutDir>
     <IntDir>Bin\Desktop_2017_Win10\$(Platform)\$(Configuration)\</IntDir>
-    <TargetName>UVAtlas</TargetName>
+    <TargetName>UVAtlasd</TargetName>
```
Comment (DillonSkeehan) [inline UVAtlas/UVAtlas_2017_Win10.vcxproj:135]: I don't think we normally commit vcxproj?

File: UVAtlas/inc/JntLinuxTypes.h [hunk 3]
```diff
@@ -0,0 +1,1653 @@
...
+//DirectXMath Port
+
+#define XM_CONST constexpr
+#define XM_CONSTEXPR constexpr
+#define XM_CALLCONV __attribute__((fastcall))
+
+namespace DirectX
+{
+
+XM_CONST float XM_PI        = 3.141592654f;
```
Comment (DillonSkeehan) [inline UVAtlas/inc/JntLinuxTypes.h:99]: may want to define these to whats included in <math.h>

File: UVAtlas/inc/UVAtlas.h [hunk 4]
```diff
@@ -127,7 +131,7 @@ namespace DirectX
     //                 maximum number of charts was too low, this gives the minimum
     //                 number of charts needed to create an atlas.
 
-    HRESULT __cdecl UVAtlasCreate(
+    HRESULT _CDECL UVAtlasCreate(
```
Comment (DillonSkeehan) [inline UVAtlas/inc/UVAtlas.h]: i guess I'd prefer if we didn't preface our custom types with _ given that it is normally reserved for system types.

File: UVAtlas/isochart/SymmetricMatrix.hpp [hunk 5]
```diff
@@ -43,7 +47,11 @@ namespace Isochart
             value_type scale,
             size_t dwDimension)
         {
+#ifndef __LINUX__
```
Comment (DillonSkeehan) [inline UVAtlas/isochart/SymmetricMatrix.hpp:50]: can we not just stud the `std` implementation in both cases?

**Review by DillonSkeehan — COMMENTED** [review]
Looks mostly okay, what files should we explicitly be looking at?

File: UVAtlas/UVAtlas_2017_Win10.vcxproj (hunk 2 above)
Comment (kiran1228) [reply UVAtlas/UVAtlas_2017_Win10.vcxproj:135]: MS originally had all these vs sln files in the repo. This was just to update the naming of the output library. I originally used this before I setup the cmake stuff.

File: CMakeLists.txt (hunk 1 above)
Comment (kiran1228) [reply CMakeLists.txt]: Not sure how to make cmake do this. Any pointers? FWIW while I'm updating dependencies I was planning on putting a note in there that the cmake isn't verified to work adding these as submodules. I can add that note to the readme too.

File: UVAtlas/inc/JntLinuxTypes.h (hunk 3 above)
Comment (kiran1228) [reply UVAtlas/inc/JntLinuxTypes.h:99]: My preference here is to stick with the same values as the MS SDK. This was a copy paste.

File: UVAtlas/inc/UVAtlas.h (hunk 4 above)
Comment (kiran1228) [reply UVAtlas/inc/UVAtlas.h]: Can change. `JNT_CDECL`?

File: UVAtlas/isochart/SymmetricMatrix.hpp (hunk 5 above)
Comment (kiran1228) [reply UVAtlas/isochart/SymmetricMatrix.hpp:50]: I was hoping to keep the MS path in tact and just patch where needed for linux.

**Comment by kiran1228** [comment]
@DillonSkeehan I would say CMake, and all the .h & .cpp except JntLinuxTypes.h & dxgiformat.h (these are just copy pastes from the SDK).

File: CMakeLists.txt (hunk 1 above)
Comment (svensht2) [reply CMakeLists.txt]: you can set it for a specific target

File: UVAtlas/inc/UVAtlas.h (hunk 4 above)
Comment (DillonSkeehan) [reply UVAtlas/inc/UVAtlas.h]: that seems better to me

File: UVAtlas/isochart/SymmetricMatrix.hpp (hunk 5 above)
Comment (DillonSkeehan) [reply UVAtlas/isochart/SymmetricMatrix.hpp:50]: fair

File: CMakeLists.txt (hunk 1 above)
Comment (kiran1228) [reply CMakeLists.txt]: So I was trying to set those flags up for a specific target but it doesn't seem to work. I think the issue is that for VS cmake has default compiler flags. Then when you use any of the target specific commands the new flags just get appended to the cl command and get ignored. It seems like the common solution is to use sub-projects and set the `CMAKE_CXX_FLAGS...` from there. They then don't set the global values but still apply to targets within in this part of the cmake tree.

**Comment by kiran1228** [comment]
Updated with @DillonSkeehan comments. If there are no red flags I think we can call this good.

**Comment by kiran1228** [comment]
merging this

**Pull request merged**